capsules/src/net/lowpan_fragment.rs

```diff
@@ -0,0 +1,900 @@
+        self.busy.set(true);
+        self.bitmap.map(|bitmap| bitmap.clear());
+        self.timeout_counter.set(0);
+    }
+
+    // This function assumes that the payload is a slice starting from the
+    // actual payload (no 802.15.4 headers, no fragmentation headers), and
+    // returns true if the packet is completely reassembled.
+    fn receive_next_frame(&self,
+                          payload: &[u8],
+                          payload_len: usize,
+                          dgram_size: u16,
+                          dgram_offset: usize,
+                          ctx_store: &ContextStore)
+                          -> Result<bool, ReturnCode> {
+        let mut packet = self.packet.take().ok_or(ReturnCode::ENOMEM)?;
+        let uncompressed_len = if dgram_offset == 0 {
+            let (consumed, written) = lowpan::decompress(ctx_store,
+                                                         &payload[0..payload_len as usize],
+                                                         self.src_mac_addr.get(),
+                                                         self.dst_mac_addr.get(),
+                                                         &mut packet,
+                                                         dgram_size,
+                                                         true).map_err(|_| ReturnCode::FAIL)?;
+            let remaining = payload_len - consumed;
+            packet[written..written + remaining]
+                .copy_from_slice(&payload[consumed..consumed + remaining]);
+            written + remaining
+
+        } else {
+            packet[dgram_offset..dgram_offset + payload_len]
+                .copy_from_slice(&payload[0..payload_len]);
+            payload_len
+        };
+        self.packet.replace(packet);
+        if !self.bitmap.map_or(false, |bitmap| {
+            bitmap.set_bits(dgram_offset / 8, (dgram_offset + uncompressed_len) / 8)
+        }) {
+            // If this fails, we received an overlapping fragment. We can simply
+            // drop the packet in this case.
+            Err(ReturnCode::FAIL)
+        } else {
+            self.bitmap
+                .map(|bitmap| bitmap.is_complete((dgram_size as usize) / 8))
+                .ok_or(ReturnCode::FAIL)
+        }
+    }
+
+    fn end_receive(&self, client: Option<&'static ReceiveClient>, result: ReturnCode) {
+        self.busy.set(false);
+        self.bitmap.map(|bitmap| bitmap.clear());
+        self.timeout_counter.set(0);
+        client.map(move |client| {
+            // Since packet is borrowed from the upper layer, failing to return it
+            // in the callback represents a significant error that should never
+            // occur - all other calls to `packet.take()` replace the packet,
+            // and thus the packet should always be here.
+            let mut buffer =
+                self.packet.take().expect("Error: `packet` is None in call to end_receive.");
```

Comment:
Next, the `packet` buffer for the `RxState` struct is owned by the `RxState` struct, passed in via the `RxState::new` function. The `expect` call on line 536 thus makes sense, as if the `packet` buffer is no longer there, then the `RxState` struct would no longer be able to receive frames, and indicates a logic bug.